src/main/java/soot/dexpler/DexFillArrayDataTransformer.java

```diff
@@ -0,0 +1,138 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ * #L%
+ */
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import soot.ArrayType;
+import soot.Body;
+import soot.BodyTransformer;
+import soot.G;
+import soot.Local;
+import soot.Type;
+import soot.Unit;
+import soot.Value;
+import soot.dexpler.instructions.FillArrayDataInstruction;
+import soot.dexpler.typing.UntypedConstant;
+import soot.jimple.ArrayRef;
+import soot.jimple.AssignStmt;
+import soot.jimple.InvokeExpr;
+import soot.jimple.NewArrayExpr;
+import soot.toolkits.graph.ExceptionalUnitGraph;
+import soot.toolkits.graph.ExceptionalUnitGraphFactory;
+import soot.toolkits.scalar.LocalDefs;
+
+/**
+ * If Dalvik bytecode can contain <code>fill-array-data</code> instructions that can fill an array with data elements we only
+ * know the element size of.
+ * 
+ * Therefore when processing such instructions in {@link FillArrayDataInstruction} we don't know the exact type of the data
+ * that is loaded. Because of (conditional) branches in the code, identifying the type is not always possible at that stage.
+ * Instead {@link UntypedConstant} constants are used. These constants are processed by this transformer and get their final
+ * type.
+ * 
+ *
+ * @author Jan Peter Stotz
+ *
+ */
+public class DexFillArrayDataTransformer extends BodyTransformer {
+  private static final Logger logger = LoggerFactory.getLogger(DexFillArrayDataTransformer.class);
+
+  public static DexFillArrayDataTransformer v() {
+    return new DexFillArrayDataTransformer();
+  }
+
+  protected void internalTransform(final Body body, String phaseName, Map<String, String> options) {
+    final ExceptionalUnitGraph g = ExceptionalUnitGraphFactory.createExceptionalUnitGraph(body, DalvikThrowAnalysis.v());
+    final LocalDefs defs = G.v().soot_toolkits_scalar_LocalDefsFactory().newLocalDefs(g);
+
+    for (Iterator<Unit> unitIt = body.getUnits().snapshotIterator(); unitIt.hasNext();) {
+      Unit u = unitIt.next();
+      if (!(u instanceof AssignStmt)) {
+        continue;
+      }
+      AssignStmt ass = (AssignStmt) u;
+      Value rightOp = ass.getRightOp();
+      if (rightOp instanceof UntypedConstant) {
+        Value left = ass.getLeftOp();
+        if (left instanceof ArrayRef) {
+          ArrayRef leftArray = (ArrayRef) left;
+
+          Local l = (Local) leftArray.getBase();
+          List<Unit> assDefs = defs.getDefsOfAt(l, ass);
+          List<Type> arrayTypes = new LinkedList<>();
+          for (Unit d : assDefs) {
+            if (d instanceof AssignStmt) {
```

Comment:
This isn't fully generic either.You might have code like this:

```
b = new Foo[];
a = b;
```

In that case, this code wouldn't be able to find the type of variable ``a``.

Reply:
Indeed this case is possible. My updated PR is now capable of handling assignment from other Local. In such a case we are tracing back the array definition using recursion. See `checkArrayDefinitions` function.